### PR TITLE
C99 conforming static assert

### PR DESCRIPTION
**Pull request by torben-hansen**
### Issues:

CryptoAlg-611

### Description of changes:

`_Static_assert` and `static_assert` are defined in C11 and does not conform to C99. This change implements a static assertion mechanism that conforms to C99. It is inspired by similar implementations in:
- AWS-C-COMMON https://github.com/awslabs/aws-c-common/blob/a61f2ea5fbc9ac2eee9fe32f7a3fdbd399d9fbdb/include/aws/common/macros.h#L16
- Perl: https://github.com/Perl/perl5/blob/5640a370e8b19af74b8ca0b4694464c21a87916b/perl.h#L3862

In addition, I attempted to improve the robustness of the C99 violation test.

Main changes are in:
- `include/openssl/type_check.h`
- `tests/coding_guidelines/c99_gcc_test.sh`

### Call-outs:

The second argument to `OPENSSL_STATIC_ASSERT` used to be a string literal. We cannot "unstringify" to construct a valid token. So, I had to replace all strings with tokens in the code base. The variable name of the typedefe'd struct should be unique to avoid shadowing already defined types. 

### Testing:

There are no run-time changes.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/openssl/type_check.h [hunk 1]
```diff
@@ -63,17 +63,24 @@
...
+// The former was used in other cases. However, these two keywords are not
+// defined before C11. So, we can't rely on these when we want to be C99
+// compliant. If we at some point decides that we want to only be compliant with
+// C11 (and up), we can reintroduce these keywords. Instead, use a method that
+// is guaranteed to be C99 compliant and still gives use an equivalent static
+// assert mechanism.
+#define AWSLC_CONCAT(left, right) left##right
+#define AWSLC_STATIC_ASSERT_DEFINE(cond, msg) typedef struct { \
+        unsigned int AWSLC_CONCAT(static_assertion_, msg) : (cond) ? 1 : - 1; \
+    } AWSLC_CONCAT(static_assertion_, msg) OPENSSL_UNUSED
```
Comment (dkostic) [inline include/openssl/type_check.h:78]: Could you please add an explanation of how this works?

And another question, did you consider using the aws-c-common solution, is there any trade-off between the two solutions?

Comment (torben-hansen) [reply include/openssl/type_check.h:78]: The width element of a bit field cannot be negative.

I guess I was easily persuaded by this line: https://github.com/Perl/perl5/blob/5640a370e8b19af74b8ca0b4694464c21a87916b/perl.h#L3860
But I don't really see why that matters. I also thought I could get around unused errors. But I couldn't, and in the end, I just stuck with the solution I had.

Comment (dkostic) [reply include/openssl/type_check.h:78]: > The width element of a bit field cannot be negative.

Thanks, but I thought maybe to add this comment in the code :)

> I guess I was easily persuaded by this line: https://github.com/Perl/perl5/blob/5640a370e8b19af74b8ca0b4694464c21a87916b/perl.h#L3860
But I don't really see why that matters. I also thought I could get around unused errors. But I couldn't, and in the end, I just stuck with the solution I had.

Thanks for the link, that explains it.

**Review by dkostic — DISMISSED** [review]

File: include/openssl/type_check.h [hunk 2]
```diff
@@ -63,17 +63,32 @@
...
+// is guaranteed to be C99 compliant and still gives use an equivalent static
+// assert mechanism.
+//
+// The solution below defines a struct type containing a bit field.
+// The name of that type is |static_assertion_msg|. |msg| is a concatenation of
+// a user-chosen error (which should be chosen with respect to actual assertion)
+// and the line the assertion is defined. This should ensure name uniqueness.
+// The width of the bit field is set to 1 or -1, depending on the evaluation of
+// the boolean expression |cond|. If the condition is false, the width requested
+// is -1, which is illegal and would cause the compiler to throw an error.
```
Comment (nebeid) [inline include/openssl/type_check.h:82]: Would it be possible add to an example of what an assertion error looks like? That, I think, would help in understanding the final output of all the macro definitions.

Comment (torben-hansen) [reply include/openssl/type_check.h:82]: Changing
```
OPENSSL_STATIC_ASSERT(sizeof(((EVP_AEAD_CTX *)NULL)->state) >=
                          sizeof(struct aead_aes_gcm_ctx),
                      AEAD_state_is_too_small)
```
to
```
OPENSSL_STATIC_ASSERT(sizeof(((EVP_AEAD_CTX *)NULL)->state) <
                          sizeof(struct aead_aes_gcm_ctx),
                      AEAD_state_is_too_small)
```
produces the following compile-time error:
```
../crypto/../include/openssl/type_check.h:85:35: error: negative width in bit-field ‘static_assertion_at_line_913_error_is_AEAD_state_is_too_small’
         unsigned int AWSLC_CONCAT(static_assertion_, msg) : (cond) ? 1 : - 1; \
                                   ^
../crypto/../include/openssl/type_check.h:83:35: note: in definition of macro ‘AWSLC_CONCAT
’
 #define AWSLC_CONCAT(left, right) left##right
                                   ^~~~
../crypto/../include/openssl/type_check.h:87:53: note: in expansion of macro ‘AWSLC_STATIC_ASSERT_DEFINE’
 #define AWSLC_STATIC_ASSERT_ADD_LINE0(cond, suffix) AWSLC_STATIC_ASSERT_DEFINE(cond, AWSLC_CONCAT(at_line_, suffix))
                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
../crypto/../include/openssl/type_check.h:88:59: note: in expansion of macro ‘AWSLC_STATIC_ASSERT_ADD_LINE0’
 #define AWSLC_STATIC_ASSERT_ADD_LINE1(cond, line, suffix) AWSLC_STATIC_ASSERT_ADD_LINE0(cond, AWSLC_CONCAT(line, suffix))
                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../crypto/../include/openssl/type_check.h:89:53: note: in expansion of macro ‘AWSLC_STATIC_ASSERT_ADD_LINE1’
 #define AWSLC_STATIC_ASSERT_ADD_LINE2(cond, suffix) AWSLC_STATIC_ASSERT_ADD_LINE1(cond, __LINE__, suffix)
                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../crypto/../include/openssl/type_check.h:90:53: note: in expansion of macro ‘AWSLC_STATIC_ASSERT_ADD_LINE2’
 #define AWSLC_STATIC_ASSERT_ADD_ERROR(cond, suffix) AWSLC_STATIC_ASSERT_ADD_LINE2(cond, AWSLC_CONCAT(_error_is_, suffix))
                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../crypto/../include/openssl/type_check.h:91:44: note: in expansion of macro ‘AWSLC_STATIC_ASSERT_ADD_ERROR’
 #define OPENSSL_STATIC_ASSERT(cond, error) AWSLC_STATIC_ASSERT_ADD_ERROR(cond, error);
                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../crypto/fipsmodule/cipher/e_aes.c:911:1: note: in expansion of macro ‘OPENSSL_STATIC_ASSERT’
 OPENSSL_STATIC_ASSERT(sizeof(((EVP_AEAD_CTX *)NULL)->state) <
```

**Review by nebeid — APPROVED** [review]

**Review by dkostic — APPROVED** [review]

**Pull request merged**